frame/contracts/src/migration/v11.rs

```diff
@@ -121,7 +121,8 @@ impl<T: Config> Migrate for Migration<T> {
 
 	#[cfg(feature = "try-runtime")]
 	fn post_upgrade_step(state: Vec<u8>) -> Result<(), TryRuntimeError> {
-		let len = <u32 as Decode>::decode(&mut &state[..]).unwrap();
+		let len = <u32 as Decode>::decode(&mut &state[..])
+			.expect("pre_upgrade_step provides a valid state; qed");
```

Comment:
all these pre/post are only use as test code, we can live with it

Reply:
> all these pre/post are only use as test code, we can live with it

yes, but how about those ones inside `Migration::step()` function?

Reply:
mmm actually that's a good point,
I wonder why we can't just do 
```diff
 pub struct Migration<T: Config> {
-	last_key: Option<BoundedVec<u8, ConstU32<256>>>,
+	last_key: Option<CodeHash<T>>,
...
-			self.last_key = Some(iter.last_raw_key().to_vec().try_into().unwrap());
+			self.last_key = Some(key);
```

to get rid of the extra unwrap
otherwise we assume that T::Hashing is no more than a H256  which in might not be necessarily true

Reply:
So I think we should go with these changes, but I will put them in a new PR
```diff
-	last_key: Option<BoundedVec<u8, ConstU32<256>>>,
+	last_key: Option<CodeHash<T>>,
 	_phantom: PhantomData<T>,
 }
 
@@ -92,7 +90,7 @@ impl<T: Config> MigrationStep for Migration<T> {
 
 	fn step(&mut self) -> (IsFinished, Weight) {
 		let mut iter = if let Some(last_key) = self.last_key.take() {
-			old::CodeStorage::<T>::iter_from(last_key.to_vec())
+			old::CodeStorage::<T>::iter_from(old::CodeStorage::<T>::hashed_key_for(last_key))
 		} else {
 			old::CodeStorage::<T>::iter()
 		};
@@ -108,7 +106,7 @@ impl<T: Config> MigrationStep for Migration<T> {
 				determinism: Determinism::Enforced,
 			};
 			CodeStorage::<T>::insert(key, module);
-			self.last_key = Some(iter.last_raw_key().to_vec().try_into().unwrap());
+			self.last_key = Some(key);
 			(IsFinished::No, T::WeightInfo::v9_migration_step(len))
 		} else {
 			log::debug!(target: LOG_TARGET, "No more contracts code to migrate");
```

Reply:
https://github.com/paritytech/substrate/pull/14418